File: api/signature_request.py
import frappe
from frappe.utils import now_datetime, add_days
from esignature.utils.document_handler import get_document_pdf, save_signed_document

@frappe.whitelist()
def send_request(signature_request):
    """Send signature request manually"""
    try:
        doc = frappe.get_doc("Signature Request", signature_request)
        
        if doc.status != "Draft":
            return {"error": "Request is not in draft status"}
        
        doc.send_signature_request_email()
        return {"success": True}
        
    except Exception as e:
        frappe.log_error(f"Manual send request failed: {str(e)}")
        return {"error": str(e)}

@frappe.whitelist()
def send_reminder(signature_request):
    """Send reminder email for pending signature"""
    try:
        doc = frappe.get_doc("Signature Request", signature_request)
        
        if doc.status not in ["Sent", "Viewed"]:
            return {"error": "Cannot send reminder for this status"}
        
        # Check if not expired
        if get_datetime(doc.expires_on) < now_datetime():
            return {"error": "Request has expired"}
        
        # Send reminder email
        signing_url = f"{frappe.utils.get_url()}/sign_document?token={doc.signature_token}"
        
        frappe.sendmail(
            recipients=[doc.recipient_email],
            subject=f"Reminder: {doc.title}",
            message=f"""
            Dear {doc.recipient_name},
            
            This is a reminder that you have a pending document to sign: {doc.title}
            
            Please click the link below to review and sign:
            {signing_url}
            
            This request expires on: {doc.expires_on}
            
            Thank you.
            """,
            reference_doctype=doc.doctype,
            reference_name=doc.name
        )
        
        return {"success": True}
        
    except Exception as e:
        frappe.log_error(f"Send reminder failed: {str(e)}")
        return {"error": str(e)}

@frappe.whitelist()
def cancel_request(signature_request):
    """Cancel signature request"""
    try:
        doc = frappe.get_doc("Signature Request", signature_request)
        
        if doc.status in ["Signed", "Cancelled"]:
            return {"error": "Cannot cancel this request"}
        
        doc.status = "Cancelled"
        doc.save()
        
        # Send cancellation email
        frappe.sendmail(
            recipients=[doc.recipient_email],
            subject=f"Cancelled: {doc.title}",
            message=f"""
            Dear {doc.recipient_name},
            
            The signature request for "{doc.title}" has been cancelled.
            
            No further action is required from your side.
            
            Thank you.
            """,
            reference_doctype=doc.doctype,
            reference_name=doc.name
        )
        
        return {"success": True}
        
    except Exception as e:
        frappe.log_error(f"Cancel request failed: {str(e)}")
        return {"error": str(e)}

@frappe.whitelist()
def resend_request(signature_request):
    """Resend expired signature request"""
    try:
        doc = frappe.get_doc("Signature Request", signature_request)
        
        if doc.status != "Expired":
            return {"error": "Can only resend expired requests"}
        
        # Reset status and extend expiry
        doc.status = "Draft"
        doc.expires_on = add_days(now_datetime(), 7)
        doc.save()
        
        # Send new request
        doc.send_signature_request_email()
        
        return {"success": True}
        
    except Exception as e:
        frappe.log_error(f"Resend request failed: {str(e)}")
        return {"error": str(e)}

@frappe.whitelist()
def get_signature_status(reference_doctype, reference_name):
    """Get signature status for a document"""
    try:
        signature_requests = frappe.get_all("Signature Request",
            filters={
                "reference_doctype": reference_doctype,
                "reference_name": reference_name
            },
            fields=["name", "status", "signed_at", "recipient_name"],
            order_by="creation desc"
        )
        
        return {"requests": signature_requests}
        
    except Exception as e:
        frappe.log_error(f"Get signature status failed: {str(e)}")
        return {"error": str(e)}

File: public/js/signature_request.js
frappe.ui.form.on('Signature Request', {
    refresh: function(frm) {
        // Add custom buttons
        if (frm.doc.status === 'Draft') {
            frm.add_custom_button(__('Send Request'), function() {
                send_signature_request(frm);
            }, __('Actions'));
        }
        
        if (frm.doc.status === 'Sent' || frm.doc.status === 'Viewed') {
            frm.add_custom_button(__('Send Reminder'), function() {
                send_reminder(frm);
            }, __('Actions'));
            
            frm.add_custom_button(__('Cancel Request'), function() {
                cancel_signature_request(frm);
            }, __('Actions'));
        }
        
        if (frm.doc.status === 'Expired') {
            frm.add_custom_button(__('Resend Request'), function() {
                resend_signature_request(frm);
            }, __('Actions'));
        }
        
        if (frm.doc.status === 'Signed' && frm.doc.signed_document_url) {
            frm.add_custom_button(__('Download Signed Document'), function() {
                window.open(frm.doc.signed_document_url, '_blank');
            }, __('View'));
        }
        
        // Add signature link button if active
        if (frm.doc.signature_token && ['Sent', 'Viewed'].includes(frm.doc.status)) {
            frm.add_custom_button(__('Copy Signature Link'), function() {
                copy_signature_link(frm);
            }, __('View'));
        }
        
        // Set indicator colors
        set_status_indicator(frm);
    },
    
    reference_doctype: function(frm) {
        if (frm.doc.reference_doctype && frm.doc.reference_name) {
            // Auto-populate recipient details from reference document
            populate_recipient_details(frm);
        }
    },
    
    reference_name: function(frm) {
        if (frm.doc.reference_doctype && frm.doc.reference_name) {
            populate_recipient_details(frm);
        }
    }
});

function send_signature_request(frm) {
    frappe.confirm(__('Send signature request to {0}?', [frm.doc.recipient_email]), function() {
        frappe.call({
            method: 'esignature.api.signature_request.send_request',
            args: {
                'signature_request': frm.doc.name
            },
            callback: function(r) {
                if (r.message && r.message.success) {
                    frappe.msgprint(__('Signature request sent successfully'));
                    frm.reload_doc();
                } else {
                    frappe.msgprint(__('Failed to send signature request'));
                }
            }
        });
    });
}

function send_reminder(frm) {
    frappe.call({
        method: 'esignature.api.signature_request.send_reminder',
        args: {
            'signature_request': frm.doc.name
        },
        callback: function(r) {
            if (r.message && r.message.success) {
                frappe.msgprint(__('Reminder sent successfully'));
            } else {
                frappe.msgprint(__('Failed to send reminder'));
            }
        }
    });
}

function cancel_signature_request(frm) {
    frappe.confirm(__('Cancel this signature request?'), function() {
        frappe.call({
            method: 'esignature.api.signature_request.cancel_request',
            args: {
                'signature_request': frm.doc.name
            },
            callback: function(r) {
                if (r.message && r.message.success) {
                    frappe.msgprint(__('Signature request cancelled'));
                    frm.reload_doc();
                }
            }
        });
    });
}

function resend_signature_request(frm) {
    frappe.confirm(__('Resend signature request?'), function() {
        frappe.call({
            method: 'esignature.api.signature_request.resend_request',
            args: {
                'signature_request': frm.doc.name
            },
            callback: function(r) {
                if (r.message && r.message.success) {
                    frappe.msgprint(__('Signature request resent successfully'));
                    frm.reload_doc();
                }
            }
        });
    });
}

function copy_signature_link(frm) {
    const signing_url = `${window.location.origin}/sign_document?token=${frm.doc.signature_token}`;
    
    if (navigator.clipboard) {
        navigator.clipboard.writeText(signing_url).then(function() {
            frappe.msgprint(__('Signature link copied to clipboard'));
        });
    } else {
        // Fallback for older browsers
        const textArea = document.createElement('textarea');
        textArea.value = signing_url;
        document.body.appendChild(textArea);
        textArea.select();
        document.execCommand('copy');
        document.body.removeChild(textArea);
        frappe.msgprint(__('Signature link copied to clipboard'));
    }
}

function populate_recipient_details(frm) {
    frappe.call({
        method: 'esignature.utils.document_handler.get_recipient_details',
        args: {
            'doctype': frm.doc.reference_doctype,
            'name': frm.doc.reference_name
        },
        callback: function(r) {
            if (r.message) {
                frm.set_value('recipient_email', r.message.email);
                frm.set_value('recipient_name', r.message.name);
                frm.set_value('title', `${frm.doc.reference_doctype} ${frm.doc.reference_name} - Signature Required`);
            }
        }
    });
}

function set_status_indicator(frm) {
    const status_colors = {
        'Draft': 'gray',
        'Sent': 'blue',
        'Viewed': 'orange', 
        'Signed': 'green',
        'Expired': 'red',
        'Cancelled': 'red'
    };
    
    frm.page.set_indicator(frm.doc.status, status_colors[frm.doc.status] || 'gray');
}
